pkg/kube-metrics/uclient.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubemetrics
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+)
+
+type Client struct {
```

Comment:
Can `Client` and `NewClientForConfig` be unexported? It looks like the only place it's used in the exported API is in `NewCollectors`, where we could just have users pass `*rest.Config` instead?

Reply:
We can, but I think these functions would be useful for the user who wants to create similar metrics or use the unstructured client on their own. WDYT?

Reply:
You're saying there are cases where users would use this `Client` in other ways outside of calling `NewCollectors()`?

Looking at this a bit closer, `NewCollectors` uses `uc` to call `ClientFor` in a loop with the namespace, but `ClientFor` doesn't actually use the namespace.  So if I understand correctly, we could actually just pass `dynamic.NamespaceableResourceInterface` to `NewCollectors` in place of `Client` (which incidentally makes it more flexible for users to pass their own implementations and easier to unit test).

Then we could refactor the `Client` code to create a `dynamic.NamespaceableResourceInterface` directly from the passed `rest.Config` (so maybe no need for a `Client` struct at all?) and call that from `ServeCRMetrics`.

```go
func ServeCRMetrics(cfg *rest.Config, ns []string, operatorGVKs []schema.GroupVersionKind, host string, port int32) error {
    ...
    for _, gvk := range operatorGVKs {
        dclient, err := newClientForGVK(cfg, gvk) // repurpose `client.ClientFor`?
        if err != nil {
            return fmt.Errorf("could get dynamic client ...")
        }
        collectors := NewCollectors(dclient, ns, gvk.GroupVersion().String(), gvk.Kind, metricFamilies)
    }
}
```

How does that sound?

Reply:
Does sounds great, done ✔️